### PR TITLE
[QNN] Support different qnn params between in/out tensor in leaky_relu

### PR DESCRIPTION
**Pull request by zhaoyang-star**
cc @AndrewZhaoLuo @margaretqian

### CODE REVIEW

File: src/relay/qnn/op/leaky_relu.cc [hunk 1]
```diff
@@ -82,42 +87,62 @@ Expr QnnLeakyReluCanonicalize(const Attrs& attrs, const Array<Expr>& new_args,
...
-  auto prod = FixedPointMultiply(quantized_data, fixed_point_multiplier, shift);
+  auto prod = FixedPointMultiply(requantized_expr, fixed_point_multiplier, shift);
 
+  // (1 - alpha) * zp_o
   int32_t fixed_point_multiplier_z, shift_z;
   std::tie(fixed_point_multiplier_z, shift_z) = GetFixedPointMultiplierShift(1 - alpha);
-  auto scaled_z = FixedPointMultiply(input_zero_point, fixed_point_multiplier_z, shift_z);
+  auto scaled_z = FixedPointMultiply(output_zero_point, fixed_point_multiplier_z, shift_z);
 
+  // alpha * Q_i + (1 - alpha) * zp_o
```
Comment (AndrewZhaoLuo) [inline src/relay/qnn/op/leaky_relu.cc:129]: // alpha * Q_i' + (1 - alpha) * zp_o

File: src/relay/qnn/op/leaky_relu.cc [hunk 2]
```diff
@@ -82,42 +87,62 @@ Expr QnnLeakyReluCanonicalize(const Attrs& attrs, const Array<Expr>& new_args,
...
   auto add = Add(prod, scaled_z);
-  auto output = Where(Less(quantized_data, input_zero_point), add, quantized_data);
+  auto output = Where(Less(data, input_zero_point), add, data);
```
Comment (AndrewZhaoLuo) [inline src/relay/qnn/op/leaky_relu.cc:131]: So `add` will be have quantization parameters of `output_scale`, `output_zero_point` which is right while `data` will have quantization parameters of `input_scale` and `input_zero_point`. Believe you must grab from requantized_expr.

File: python/tvm/relay/qnn/op/qnn.py [hunk 3]
```diff
@@ -1192,15 +1192,13 @@ def leaky_relu(x, alpha, scale, zero_point):
         The scale of the quantized expr.
```
[the remark this answers is not in the document]
Comment (zhaoyang-star) [reply python/tvm/relay/qnn/op/qnn.py]: have fixed.

File: python/tvm/relay/qnn/op/qnn.py [hunk 4]
```diff
@@ -1179,7 +1179,7 @@ def batch_matmul(x, y, x_zero_point, y_zero_point, x_scale, y_scale, out_dtype="
 reg.register_pattern("qnn.dequantize", OpPattern.OPAQUE)
 
 
-def leaky_relu(x, alpha, scale, zero_point):
+def leaky_relu(x, alpha, scale, zero_point, output_scale, output_zero_point):
```
[the remark this answers is not in the document]
Comment (zhaoyang-star) [reply python/tvm/relay/qnn/op/qnn.py]: Done

File: src/relay/qnn/op/leaky_relu.cc [hunk 5]
```diff
@@ -82,42 +87,62 @@ Expr QnnLeakyReluCanonicalize(const Attrs& attrs, const Array<Expr>& new_args,
...
-  // = a * q_t + (1 - a) * z
+  // LeakyReLU can be written in terms of respective quantized tensors, scales and
+  // zero points as
   //
-  // We return the quantized value of alpha * T for all values q_t < input_zero_point.
-
-  ICHECK_EQ(new_args.size(), 3);
-  Expr quantized_data = Cast(new_args[0], DataType::Int(32));
+  //    scale_o * (Q_o - zp_o) = alpha * scale_i * (Q_i - zp_i)  when Q_i < zp_i  (1)
+  //                       Q_o = Q_i when Q_i >= zp_i  (2)
```
[the remark this answers is not in the document]
Comment (zhaoyang-star) [reply src/relay/qnn/op/leaky_relu.cc]: Yes, I misunderstood it when `Q_i >= zp_i`. Have fixed equation (2) .

File: src/relay/qnn/op/leaky_relu.cc [hunk 6]
```diff
@@ -82,42 +87,62 @@ Expr QnnLeakyReluCanonicalize(const Attrs& attrs, const Array<Expr>& new_args,
...
+  //
+  // Since the input qnn params can be different than output qnn params, we first requantize the
+  // input tensor to the output qnn params. After requantizing Q_i, equation (1) becames equation
+  // (3) where Q_i' is the requantized data from Q_i.
+  //
+  //    scale_o * (Q_o - zp_o) = alpha * scale_o * (Q_i' - zp_o)  when Q_i < zp_i (3)
+  //                       Q_o = alpha * Q_i' + (1 - alpha) * zp_o  when Q_i < zp_i (4)
```
[the remark this answers is not in the document]
Comment (zhaoyang-star) [reply src/relay/qnn/op/leaky_relu.cc]: Done

File: src/relay/qnn/op/leaky_relu.cc (hunk 2 above)
Comment (zhaoyang-star) [reply src/relay/qnn/op/leaky_relu.cc]: Thanks! According to equation (2) we should return `requantized_expr` rather than `data`.

**Comment by zhaoyang-star** [comment]
Please review again.
Thank you.

**Review by AndrewZhaoLuo — APPROVED** [review]

**Pull request merged**